tx_json.go

```diff
@@ -0,0 +1,457 @@
+package wavelet
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/perlin-network/wavelet/sys"
+	"github.com/pkg/errors"
+	"github.com/valyala/fastjson"
+)
+
+// Parser defines a generic JSON transaction payload parser.
+type Parser struct {
+	Tag string // Transaction tag
+}
+
+var (
+	// ErrNoTag defines an error describing an empty TransactionTag.
+	ErrNoTag = errors.New("no tag specified")
+
+	// ErrInvalidTag defines an error describing an invalid TransactionTag.
+	ErrInvalidTag = errors.New("tag is invalid")
+
+	// ErrCouldNotParse defines an error describing the inability to parse a given payload.
+	ErrCouldNotParse = errors.New("could not parse the given payload")
+
+	// ErrNilField defines an error describing a field value equal to nil.
+	ErrNilField = errors.New("field is nil")
+
+	// ErrInvalidOperation defines an error describing an invalid operation value.
+	ErrInvalidOperation = errors.New("operation is invalid")
+)
+
+/* BEGIN EXPORTED METHODS */
+
+// NewParser initializes a new parser with the given transaction type.
+func NewParser(transactionTag string) *Parser {
+	return &Parser{
+		Tag: transactionTag, // Set transaction tag
+	} // Initialize parser
+}
+
+// ParseJSON parses the given JSON payload input.
+func (parser *Parser) ParseJSON(data []byte) ([]byte, error) {
+	if parser.Tag == "" { // Check no transaction tag
+		return nil, ErrNoTag // Return no tag error
+	}
+
+	if !getValidTags()[parser.Tag] { // Check invalid
+		return nil, ErrInvalidTag // Return error
+	}
+
+	switch parser.Tag { // Handle different tag types
+	case "nop":
+		return nil, nil // Nothing to do!
+	case "transfer":
+		return parser.parseTransfer(data) // Parse
+	case "stake":
+		return parser.parseStake(data) // Parse
+	case "contract":
+		return parser.parseContract(data) // Parse
+	case "batch":
+		return parser.parseBatch(data) // Parse
+	}
+
+	return nil, ErrCouldNotParse // Return error (shouldn't ever get here)
+}
+
+/* END EXPORTED METHODS */
+
+/* BEGIN INTERNAL METHODS */
+
+/*
+	BEGIN TAG HANDLERS
+*/
+
+// parseTransfer parses a transaction payload with the transfer tag.
+func (parser *Parser) parseTransfer(data []byte) ([]byte, error) {
+	payload := bytes.NewBuffer(nil) // Initialize buffer
+
+	parsedData, err := parseJSONBytes(data) // Parse data
```

Comment:
It might be better to in-line a lot of these methods such as `parseJSONBytes` and `parseRecipient` and `parseAmount` to avoid far too much encapsulation.

Reply:
Would apply my comment above to all other functions below.

Reply:
I understand why parseJSONBytes might be inlined, but here it does feel like encapsulating parser methods for each different field provides some marginal benefit. Seems like inlining them would make things pretty messy (and a little hard to refactor later).

Reply:
Not to mention you'd be dealing with a lot of unnecessary duplicate code.